File: app/src/main/java/com/jobtick/activities/SearchCategoryActivity.java
package com.jobtick.activities;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.view.KeyEvent;
import android.view.View;
import android.view.inputmethod.EditorInfo;
import android.view.inputmethod.InputMethodManager;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.TextView;

import androidx.annotation.Nullable;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.android.volley.AuthFailureError;
import com.android.volley.DefaultRetryPolicy;
import com.android.volley.Request;
import com.android.volley.RequestQueue;
import com.android.volley.toolbox.StringRequest;
import com.android.volley.toolbox.Volley;
import com.google.android.material.button.MaterialButton;
import com.jobtick.R;
import android.annotation.SuppressLint;

import com.jobtick.adapers.TaskCategoryAdapter;
import com.jobtick.adapers.PreviewAdapter;
import com.jobtick.models.PreviewCategorySetModel;
import com.jobtick.models.TaskCategory;
import com.jobtick.models.PreviewCategoryModel;
import com.jobtick.pagination.PaginationListener;
import com.jobtick.utils.Constant;
import com.jobtick.utils.ConstantKey;
import com.jobtick.utils.SessionManager;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

import static com.jobtick.pagination.PaginationListener.PAGE_START;

public class SearchCategoryActivity extends ActivityBase implements TextView.OnEditorActionListener,
        TaskCategoryAdapter.OnItemClickListener, PreviewAdapter.OnItemClickListener<PreviewCategoryModel> {

    @SuppressLint("NonConstantResourceId")
    @BindView(R.id.lyt_search_new)
    MaterialButton lytSearchNew;

    @SuppressLint("NonConstantResourceId")
    @BindView(R.id.iv_back)
    ImageView ivBack;

    @SuppressLint("NonConstantResourceId")
    @BindView(R.id.list)
    RecyclerView recyclerViewCategories;

    @SuppressLint("NonConstantResourceId")
    @BindView(R.id.empty_search)
    RelativeLayout emptySearch;

    @SuppressLint("NonConstantResourceId")
    @BindView(R.id.edt_search_categoreis)
    EditText edtSearchCategories;

    private SessionManager sessionManager;
    private TaskCategoryAdapter adapter;
    private PreviewCategorySetModel previewCategorySetModel;
    private PreviewAdapter previewAdapter;

    private int currentPage = PAGE_START;
    private boolean isLastPage = false;
    private int totalPage = 10;
    private boolean isLoading = false;

    private String query = "";

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_search_all);
        ButterKnife.bind(this);
        //  RelativeLayout emptySearch = findViewById(R.id.empty_search);
        sessionManager = new SessionManager(this);

        edtSearchCategories.requestFocus();
        edtSearchCategories.performClick();
        edtSearchCategories.setOnEditorActionListener(this);

        setPreviewAdapter();
    }

    @SuppressLint("NonConstantResourceId")
    @OnClick({R.id.lyt_search_new, R.id.iv_back})
    public void onViewClicked(View view) {
        switch (view.getId()) {
            case R.id.lyt_search_new:
                edtSearchCategories.setText("");
                edtSearchCategories.requestFocus();
                edtSearchCategories.performClick();
                showKeyboard(edtSearchCategories);
                break;
            case R.id.iv_back:
                finish();
                break;

        }

    }

    @Override
    public boolean onEditorAction(TextView textView, int actionId, KeyEvent keyEvent) {
        if (actionId == EditorInfo.IME_ACTION_SEARCH) {
            //set adapter to online mode (previewMode will be false)
            setOnlineAdapter();
            setLoadMoreListener();
            getTaskCategoryData();
            return true;
        }
        return false;
    }


    public List<TaskCategory> getTaskCategoryData() {
        List<TaskCategory> items = new ArrayList<>();
        query = edtSearchCategories.getText().toString();
        StringRequest stringRequest = new StringRequest(Request.Method.GET, Constant.BASE_URL + Constant.TASK_CATEGORY + "?query=" + query + "&page=" + currentPage,
                response -> {
                    try {
                        JSONObject jsonObject = new JSONObject(response);
                        if (jsonObject.has("data") && !jsonObject.isNull("data")) {
                            JSONArray jsonArray_data = jsonObject.getJSONArray("data");
                            for (int i = 0; jsonArray_data.length() > i; i++) {
                                JSONObject jsonObject_taskModel_list = jsonArray_data.getJSONObject(i);
                                TaskCategory taskModel = new TaskCategory().getJsonToModel(jsonObject_taskModel_list, SearchCategoryActivity.this);
                                items.add(taskModel);
                            }
                        } else {
                            showToast("some went to wrong", SearchCategoryActivity.this);
                            return;
                        }

                        if (jsonObject.has("meta") && !jsonObject.isNull("meta")) {
                            JSONObject jsonObject_meta = jsonObject.getJSONObject("meta");
                            totalPage = jsonObject_meta.getInt("last_page");
                            Constant.PAGE_SIZE = jsonObject_meta.getInt("per_page");
                        }

                        if (currentPage != PAGE_START)
                            adapter.removeLoading();
                        if (items.size() <= 0) {
                            emptySearch.setVisibility(View.VISIBLE);
                            recyclerViewCategories.setVisibility(View.GONE);
                        } else {
                            recyclerViewCategories.setVisibility(View.VISIBLE);
                            emptySearch.setVisibility(View.GONE);
                        }
                        adapter.addItems(items);

                        if (currentPage < totalPage) {
                            adapter.addLoading();
                        } else {
                            isLastPage = true;
                        }
                        isLoading = false;
                    } catch (JSONException e) {
                        hideProgressDialog();
                        e.printStackTrace();
                    }
                },
                error -> errorHandle1(error.networkResponse)) {
            @Override
            public Map<String, String> getHeaders() throws AuthFailureError {
                Map<String, String> map1 = new HashMap<String, String>();
                map1.put("Content-Type", "application/x-www-form-urlencoded");
                map1.put("Authorization", "Bearer " + sessionManager.getAccessToken());
                return map1;
            }
        };

        stringRequest.setRetryPolicy(new DefaultRetryPolicy(0, -1,
                DefaultRetryPolicy.DEFAULT_BACKOFF_MULT));
        RequestQueue requestQueue = Volley.newRequestQueue(SearchCategoryActivity.this);
        requestQueue.add(stringRequest);
        return items;
    }

    private void setPreviewAdapter() {
        previewCategorySetModel = sessionManager.getPreviewModel(SearchCategoryActivity.class);
        if (previewCategorySetModel == null)
            previewCategorySetModel = new PreviewCategorySetModel();

        previewAdapter = new PreviewAdapter(new ArrayList<>(previewCategorySetModel.getPreviewSet()));
        recyclerViewCategories.setAdapter(previewAdapter);

        recyclerViewCategories.setHasFixedSize(true);
        previewAdapter.setOnItemClickListener(this);
    }

    private void setOnlineAdapter(){
        adapter = new TaskCategoryAdapter(this, new ArrayList<>());
        recyclerViewCategories.setAdapter(adapter);
        adapter.setOnItemClickListener(this);
    }

    private boolean previewMode = true;
    private void setLoadMoreListener() {
        if (!previewMode) return;
        recyclerViewCategories.addOnScrollListener(new PaginationListener((LinearLayoutManager) recyclerViewCategories.getLayoutManager()) {
            @Override
            protected void loadMoreItems() {
                isLoading = true;
                currentPage++;
                getTaskCategoryData();
            }

            @Override
            public boolean isLastPage() {
                return isLastPage;
            }

            @Override
            public boolean isLoading() {
                return isLoading;
            }
        });
        previewMode = false;
    }

    private void showKeyboard(EditText editText) {
        editText.post(new Runnable() {
            @Override
            public void run() {
                InputMethodManager imm = (InputMethodManager) getBaseContext()
                        .getSystemService(Context.INPUT_METHOD_SERVICE);
                imm.showSoftInput(editText, 0);
            }
        });
    }

    @Override
    public void onItemClick(View view, PreviewCategoryModel obj, int position) {
        previewCategorySetModel.addItem(obj);
        sessionManager.setPreviewModel(previewCategorySetModel, SearchCategoryActivity.class);
        Intent creating_task = new Intent(SearchCategoryActivity.this, TaskCreateActivity.class);
        Bundle bundle = new Bundle();
        bundle.putInt(ConstantKey.CATEGORY_ID, obj.getId());
        creating_task.putExtras(bundle);
        startActivityForResult(creating_task, ConstantKey.RESULTCODE_CATEGORY);
        finish();
    }

    @Override
    public void onItemClick(View view, TaskCategory obj, int position) {
        onItemClick(view, obj.getPreviewModel(),position);
    }
}

File: app/src/main/java/com/jobtick/android/activities/SearchCategoryActivity.kt
package com.jobtick.android.activities

import android.content.Intent
import android.os.Bundle
import android.view.KeyEvent
import android.view.View
import android.view.inputmethod.EditorInfo
import android.view.inputmethod.InputMethodManager
import android.widget.EditText
import android.widget.ImageView
import android.widget.RelativeLayout
import android.widget.TextView
import android.widget.TextView.OnEditorActionListener
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.android.volley.*
import com.android.volley.toolbox.StringRequest
import com.android.volley.toolbox.Volley
import com.google.android.material.button.MaterialButton
import com.jobtick.android.BuildConfig
import com.jobtick.android.R
import com.jobtick.android.adapers.PreviewAdapter
import com.jobtick.android.adapers.TaskCategoryAdapter
import com.jobtick.android.models.PreviewCategoryModel
import com.jobtick.android.models.PreviewCategorySetModel
import com.jobtick.android.models.TaskCategory
import com.jobtick.android.pagination.PaginationListener
import com.jobtick.android.utils.Constant
import com.jobtick.android.utils.ConstantKey
import com.jobtick.android.utils.SessionManager
import org.json.JSONException
import org.json.JSONObject
import java.util.*

class SearchCategoryActivity : ActivityBase(), OnEditorActionListener, TaskCategoryAdapter.OnItemClickListener, PreviewAdapter.OnItemClickListener<PreviewCategoryModel?> {
    private lateinit var lytSearchNew: MaterialButton
    private lateinit var ivBack: ImageView
    private lateinit var recyclerViewCategories: RecyclerView
    private lateinit var emptySearch: RelativeLayout
    private lateinit var edtSearchCategories: EditText
    private var sessionManagerS: SessionManager? = null
    private var adapter: TaskCategoryAdapter? = null
    private var previewCategorySetModel: PreviewCategorySetModel? = null
    private var previewAdapter: PreviewAdapter? = null
    private var currentPage = PaginationListener.PAGE_START
    private var isLastPageItem = false
    private var totalPage = 10
    private var isLoadingItem = false
    private var query = ""
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_search_all)
        setIDs()
        onViewClick()
        //  RelativeLayout emptySearch = findViewById(R.id.empty_search);
        sessionManagerS = SessionManager(this)
        edtSearchCategories.requestFocus()
        edtSearchCategories.performClick()
        edtSearchCategories.setOnEditorActionListener(this)
        setPreviewAdapter()
    }

    private fun setIDs() {
        lytSearchNew = findViewById(R.id.lyt_search_new)
        lytSearchNew = findViewById(R.id.iv_back)
        recyclerViewCategories = findViewById(R.id.list)
        emptySearch = findViewById(R.id.empty_search)
        edtSearchCategories = findViewById(R.id.edt_search_categoreis)
    }

   private fun onViewClick() {
        ivBack.setOnClickListener {
            onBackPressed()
        }
        lytSearchNew.setOnClickListener {
            edtSearchCategories.setText("")
            edtSearchCategories.requestFocus()
            edtSearchCategories.performClick()
            showKeyboard(edtSearchCategories)
        }

    }

    override fun onEditorAction(textView: TextView?, actionId: Int, keyEvent: KeyEvent?): Boolean {
        if (actionId == EditorInfo.IME_ACTION_SEARCH) {
            //set adapter to online mode (previewMode will be false)
            setOnlineAdapter()
            setLoadMoreListener()
            taskCategoryData
            return true
        }
        return false
    }

    val taskCategoryData: List<TaskCategory>
        get() {
            val items: MutableList<TaskCategory> = ArrayList()
            query = edtSearchCategories.text.toString()
            val stringRequest: StringRequest = object : StringRequest(Method.GET,
                    Constant.BASE_URL + Constant.TASK_CATEGORY + "?query=" + query + "&page=" + currentPage,
                    Response.Listener { response: String? ->
                        try {
                            val jsonObject = JSONObject(response!!)
                            if (jsonObject.has("data") && !jsonObject.isNull("data")) {
                                val jsonArrayData = jsonObject.getJSONArray("data")
                                var i = 0
                                while (jsonArrayData.length() > i) {
                                    val jsonobjectTaskmodelList = jsonArrayData.getJSONObject(i)
                                    val taskModel = TaskCategory().getJsonToModel(jsonobjectTaskmodelList, this@SearchCategoryActivity)
                                    items.add(taskModel)
                                    i++
                                }
                            } else {
                                showToast("some went to wrong", this@SearchCategoryActivity)
                                return@Listener
                            }
                            if (jsonObject.has("meta") && !jsonObject.isNull("meta")) {
                                val jsonObjectMeta = jsonObject.getJSONObject("meta")
                                totalPage = jsonObjectMeta.getInt("last_page")
                                Constant.PAGE_SIZE = jsonObjectMeta.getInt("per_page")
                            }
                            if (currentPage != PaginationListener.PAGE_START) adapter!!.removeLoading()
                            if (items.size <= 0) {
                                emptySearch.visibility = View.VISIBLE
                                recyclerViewCategories.visibility = View.GONE
                            } else {
                                recyclerViewCategories.visibility = View.VISIBLE
                                emptySearch.visibility = View.GONE
                            }
                            adapter!!.addItems(items)
                            if (currentPage < totalPage) {
                                adapter!!.addLoading()
                            } else {
                                isLastPageItem = true
                            }
                            isLoadingItem = false
                        } catch (e: JSONException) {
                            hideProgressDialog()
                            e.printStackTrace()
                        }
                    },
                    Response.ErrorListener { error: VolleyError -> errorHandle1(error.networkResponse) }) {
                override fun getHeaders(): Map<String, String> {
                    val map1: MutableMap<String, String> = HashMap()
                    map1["Content-Type"] = "application/x-www-form-urlencoded"
                    map1["Authorization"] = "Bearer " + sessionManagerS!!.accessToken
                    map1["Version"] = BuildConfig.VERSION_CODE.toString()
                    return map1
                }
            }
            stringRequest.retryPolicy = DefaultRetryPolicy(0, -1,
                    DefaultRetryPolicy.DEFAULT_BACKOFF_MULT)
            val requestQueue = Volley.newRequestQueue(this@SearchCategoryActivity)
            requestQueue.add(stringRequest)
            return items
        }

    private fun setPreviewAdapter() {
        previewCategorySetModel = sessionManagerS!!.getPreviewModel(SearchCategoryActivity::class.java)
        if (previewCategorySetModel == null) previewCategorySetModel = PreviewCategorySetModel()
        previewAdapter = PreviewAdapter(ArrayList(previewCategorySetModel!!.previewSet))
        recyclerViewCategories.adapter = previewAdapter
        recyclerViewCategories.setHasFixedSize(true)
        previewAdapter!!.setOnItemClickListener(this)
    }

    private fun setOnlineAdapter() {
        adapter = TaskCategoryAdapter(this, ArrayList())
        recyclerViewCategories.adapter = adapter
        adapter!!.setOnItemClickListener(this)
    }

    private var previewMode = true
    private fun setLoadMoreListener() {
        if (!previewMode) return
        recyclerViewCategories.addOnScrollListener(object : PaginationListener((recyclerViewCategories.layoutManager as LinearLayoutManager?)!!) {
            override fun loadMoreItems() {
                isLoadingItem = true
                currentPage++
                taskCategoryData
            }

            override val isLastPage: Boolean
                get() = isLastPageItem
            override val isLoading: Boolean
                get() = isLoadingItem
        })
        previewMode = false
    }

    private fun showKeyboard(editText: EditText?) {
        editText!!.post {
            val imm = baseContext
                    .getSystemService(INPUT_METHOD_SERVICE) as InputMethodManager
            imm.showSoftInput(editText, 0)
        }
    }

    override fun onItemClick(view: View, obj: PreviewCategoryModel?, position: Int) {
        previewCategorySetModel!!.addItem(obj)
        sessionManagerS!!.setPreviewModel(previewCategorySetModel, SearchCategoryActivity::class.java)
        val creatingTask = Intent(this@SearchCategoryActivity, TaskCreateActivity::class.java)
        val bundle = Bundle()
        bundle.putInt(ConstantKey.CATEGORY_ID, obj!!.id)
        creatingTask.putExtras(bundle)
        startActivityForResult(creatingTask, ConstantKey.RESULTCODE_CATEGORY)
        finish()
    }

    override fun onItemClick(view: View, obj: TaskCategory, position: Int) {
        onItemClick(view, obj.previewModel, position)
    }
}